Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,579 @@
+  containing `E`, such that the map `(E →+* L) → (F →+* L)` induced by `algebraMap F E` is
+  injective, then `E / F` is purely inseparable. In other words, epimorphisms in the category of
+  fields must be purely inseparable extensions. Need to use the fact that `Emb F E` is infintie
+  when `E / F` is (purely) transcendental.
+
+- Prove that the (infinite) separable degree multiply by the (infinite) inseparable degree is equal
+  to the (infinite) field extension degree. Need to use the linearly disjoint argument.
+
+-/
+
+open scoped Classical Polynomial
+
+open FiniteDimensional Polynomial IntermediateField Field
+
+noncomputable section
+
+universe u v w
+
+variable (F : Type u) (E : Type v) [Field F] [Field E] [Algebra F E]
+
+variable (K : Type w) [Field K] [Algebra F K]
+
+section IsPurelyInseparable
+
+/-- Typeclass for purely inseparable field extension: an algebraic extension `E / F` is purely
+inseparable if and only if the minimal polynomial of every element of `E ∖ F` is not separable. -/
+class IsPurelyInseparable : Prop where
+  isIntegral' (x : E) : IsIntegral F x
+  inseparable' (x : E) : (minpoly F x).Separable → x ∈ (algebraMap F E).range
+
+theorem IsPurelyInseparable.isAlgebraic [IsPurelyInseparable F E] :
+    Algebra.IsAlgebraic F E := fun x ↦ (IsPurelyInseparable.isIntegral' x).isAlgebraic
+
+variable {E}
+
+theorem IsPurelyInseparable.isIntegral [IsPurelyInseparable F E] : Algebra.IsIntegral F E :=
+  IsPurelyInseparable.isIntegral'
+
+theorem IsPurelyInseparable.inseparable [IsPurelyInseparable F E] :
+    ∀ x : E, (minpoly F x).Separable → x ∈ (algebraMap F E).range :=
+  IsPurelyInseparable.inseparable'
+
+variable {F K}
+
+theorem isPurelyInseparable_iff : IsPurelyInseparable F E ↔ ∀ x : E,
+    IsIntegral F x ∧ ((minpoly F x).Separable → x ∈ (algebraMap F E).range) :=
+  ⟨fun h x ↦ ⟨h.isIntegral' x, h.inseparable' x⟩, fun h ↦ ⟨fun x ↦ (h x).1, fun x ↦ (h x).2⟩⟩
+
+/-- Transfer `IsPurelyInseparable` across an `AlgEquiv`. -/
+theorem AlgEquiv.isPurelyInseparable (e : K ≃ₐ[F] E) [IsPurelyInseparable F K] :
+    IsPurelyInseparable F E := by
+  refine ⟨fun _ ↦ by rw [← isIntegral_algEquiv e.symm]; exact IsPurelyInseparable.isIntegral F _,
+    fun x h ↦ ?_⟩
+  rw [← minpoly.algEquiv_eq e.symm] at h
+  simpa only [RingHom.mem_range, algebraMap_eq_apply] using IsPurelyInseparable.inseparable F _ h
+
+theorem AlgEquiv.isPurelyInseparable_iff (e : K ≃ₐ[F] E) :
+    IsPurelyInseparable F K ↔ IsPurelyInseparable F E :=
+  ⟨fun _ ↦ e.isPurelyInseparable, fun _ ↦ e.symm.isPurelyInseparable⟩
+
+/-- If `E / F` is an algebraic extension, `F` is separably closed,
+then `E / F` is purely inseparable. -/
+theorem Algebra.IsAlgebraic.isPurelyInseparable_of_isSepClosed (halg : Algebra.IsAlgebraic F E)
+    [IsSepClosed F] : IsPurelyInseparable F E :=
+  ⟨halg.isIntegral, fun x h ↦ minpoly.mem_range_of_degree_eq_one F x <|
+    IsSepClosed.degree_eq_one_of_irreducible F (minpoly.irreducible (halg x).isIntegral) h⟩
+
+variable (F E K)
```

Comment:
I think some lemmas below should have implicit variables, for example `eq_bot_of_isPurelyInseparable_of_isSeparable` or `isPurelyInseparable_iff_natSepDegree_eq_one `

Reply:
Done.